### PR TITLE
Contacts UI

### PR DESCRIPTION
**Pull request by sorja**
- [x] Use new grid
- [x] Check types / data passed for components
- [x] Remove old deprecated component usage (textinput, select)

### CODE REVIEW

File: src/client/components/TableNodeExt/CellNodeExt/CellNodeExt.tsx [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+const CellNodeExt: React.FC<Props> = (props: Props) => {
+  const { column, datum, disabled, lastRow, onChange, uuid, lastCol } = props
+
+  const { type } = column
+  const { colName } = column.props
+  const Component = components[type]
+
+  return (
+    <DataCell lastCol={lastCol} lastRow={lastRow}>
```
Comment (sorja) [inline src/client/components/TableNodeExt/CellNodeExt/CellNodeExt.tsx:37]: Added `lastCol`, `lastRow` prop and use new `DataCell`

File: src/client/components/TableNodeExt/CellNodeExt/CellSelect/CellSelect.tsx [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+
+import { CellProps } from '../CellProps'
+import { useValues } from './hooks/useValues'
+
+const CellSelect: React.FC<CellProps & { value: string }> = (props: CellProps & { value: string }) => {
+  const { onChange, disabled, column, value: _value } = props
+
+  const { value, options } = useValues({ column, value: _value })
+
+  return <Select disabled={disabled} value={value} onChange={onChange} options={options} />
```
Comment (sorja) [inline src/client/components/TableNodeExt/CellNodeExt/CellSelect/CellSelect.tsx:13]: Introduced hook `useValues` and use new `Select` component

File: src/client/components/TableNodeExt/CellNodeExt/CellSelect/hooks/useValues.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+          })
+        ),
+      })),
+    [options, t]
+  )
+  const _value = useMemo(() => _options.find((option) => option.value === value), [_options, value])
+
+  return {
+    value: _value,
+    options: _options,
```
Comment (sorja) [inline src/client/components/TableNodeExt/CellNodeExt/CellSelect/hooks/useValues.ts:29]: New hook to return `value` and `options` for the `CellSelect`

File: src/client/components/TableNodeExt/CellNodeExt/CellText/CellText.tsx [hunk 4]
```diff
@@ -0,0 +1,13 @@
+import React from 'react'
+
+import TextInput from 'client/components/Inputs/InputText'
+
+import { CellProps } from '../CellProps'
+
+const CellText: React.FC<CellProps & { value: string }> = (props: CellProps & { value: string }) => {
+  const { value, onChange, disabled } = props
+
+  return <TextInput disabled={disabled} value={value} onChange={(event) => onChange(event.target.value)} />
```
Comment (sorja) [inline src/client/components/TableNodeExt/CellNodeExt/CellText/CellText.tsx:10]: Only changed to new `TextInput` element

File: src/client/components/TableNodeExt/TableNodeExt.tsx [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+  disabled: boolean
+  onChange: (uuid: string, colName: string, value: any) => void
+}
+
+const TableNodeExt = (props: Props) => {
+  const { columns, data, onChange, disabled } = props
+  const { t } = useTranslation()
+
+  return (
+    <DataGrid gridTemplateColumns={`repeat(${columns.length}, auto)`}>
```
Comment (sorja) [inline src/client/components/TableNodeExt/TableNodeExt.tsx:24]: Use new `Datagrid`

File: src/client/components/TableNodeExt/TableNodeExt.tsx [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+      {columns.map((column, i) => {
+        const { colName, header } = column.props
+        return (
+          <DataCell lastCol={i === columns.length - 1} header key={`${colName}_header`}>
```
Comment (sorja) [inline src/client/components/TableNodeExt/TableNodeExt.tsx:28]: Use new `DataCell` with props `lastCol`

File: src/client/components/TableNodeExt/TableNodeExt.tsx [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+              return (
+                <CellNodeExt
+                  uuid={uuid}
+                  key={key}
+                  onChange={onChange}
+                  disabled={disabled}
+                  datum={datum}
+                  column={column}
+                  lastRow={i === data.length - 1}
+                  lastCol={j === columns.length - 1}
```
Comment (sorja) [inline src/client/components/TableNodeExt/TableNodeExt.tsx:49]: Pass new prop `lastRow` and `lastCol` which are used for `DataCell` inside of `CellNodeExt`

File: src/client/components/TableNodeExt/types.ts [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+
+import { Option } from 'client/components/Inputs/Select/types'
+
+export type ColumnNodeExt = {
+  type: ColumnNodeExtType
+  props: {
+    colName: string
+    header: string
+    // label is passed as Label, but translated before rendering
+    options?: Array<Omit<Option, 'label'> & { label: Label }>
```
Comment (sorja) [inline src/client/components/TableNodeExt/types.ts]: Updated type to support changing language

Comment (sorja) [reply src/client/components/TableNodeExt/types.ts]: Translated in `useValues` hook 

```
// useValues.ts
  const _options = useMemo(
    () =>
      options.map((option) => ({
        ...option,
        label: t(
          Labels.getLabel({
            label: option.label,
            t,
          })
        ),
      })),
    [options, t]
  )
```

File: src/client/components/TableNodeExt/CellNodeExt/CellMultiselect/CellMultiselect.tsx [hunk 9]
```diff
@@ -0,0 +1,10 @@
+import React from 'react'
+
+import { CellProps } from '../CellProps'
+
+const CellMultiselect: React.FC<CellProps & { value: string }> = (props: CellProps & { value: string }) => {
```
Comment (minotogna) [inline src/client/components/TableNodeExt/CellNodeExt/CellMultiselect/CellMultiselect.tsx]: maybe value is a prop of CellProps already ?

Comment (minotogna) [reply src/client/components/TableNodeExt/CellNodeExt/CellMultiselect/CellMultiselect.tsx]: should the value here, in this case be an array of string ? to cast internally to component, don't override prop

File: src/client/components/TableNodeExt/types.ts (hunk 8 above)
Comment (minotogna) [inline src/client/components/TableNodeExt/types.ts]: should options be a sub property of select ?

File: src/client/components/TableNodeExt/CellNodeExt/CellMultiselect/CellMultiselect.tsx (hunk 9 above)
Comment (sorja) [reply src/client/components/TableNodeExt/CellNodeExt/CellMultiselect/CellMultiselect.tsx]: You are right, it was already defined as `string | Array<string>`

```
export type CellProps = {
  value: string | Array<string>
  onChange: (newValue: string) => void
  disabled: boolean
  column: ColumnNodeExt
}

```

In the components we have to override the type though, eg.
`const CellMultiselect: React.FC<CellProps & { value: Array<string> }> = (`

So we know what kind of value this component is handling

I updated it as above, to `Array<string>`

File: src/client/components/TableNodeExt/types.ts (hunk 8 above)
Comment (sorja) [reply src/client/components/TableNodeExt/types.ts]: Currently it is defined here:

```
  const columns: Array<ColumnNodeExt> = [
    {
      type: ColumnNodeExtType.select,
      props: { colName: 'appellation', header: 'editUser.title', options: optionsAppellation },
    },
    { type: ColumnNodeExtType.text, props: { colName: 'name', header: 'editUser.name' } },
    { type: ColumnNodeExtType.text, props: { colName: 'surname', header: 'editUser.surname' } },
    { type: ColumnNodeExtType.select, props: { colName: 'role', header: 'editUser.role', options: optionsRole } },
    { type: ColumnNodeExtType.text, props: { colName: 'institution', header: 'editUser.institution' } },
    { type: ColumnNodeExtType.multiselect, props: { colName: 'contributions', header: 'editUser.contributions' } },
  ]
```

You mean we move it one level higher, or what?

Comment (minotogna) [reply src/client/components/TableNodeExt/types.ts]: I mean one level down. see col.ts. to organize props in a more consistent way. what do you think?

Comment (minotogna) [reply src/client/components/TableNodeExt/types.ts]: also, I'm look at header. which is the header label, right? then maybe organize it like 
```
header:{
label:Label
}
```
what do you think?

File: src/client/components/TableNodeExt/CellNodeExt/CellMultiselect/CellMultiselect.tsx [hunk 10]
```diff
@@ -0,0 +1,12 @@
+import React from 'react'
+
+import { CellProps } from '../CellProps'
+
+const CellMultiselect: React.FC<CellProps & { value: Array<string> }> = (
+  props: CellProps & { value: Array<string> }
+) => {
```
Comment (minotogna) [inline src/client/components/TableNodeExt/CellNodeExt/CellMultiselect/CellMultiselect.tsx]: not sure it's nice to see the override of value duplicated twice here.
either cast within the component or redefine Props above

File: src/client/components/TableNodeExt/types.ts (hunk 8 above)
Comment (minotogna) [reply src/client/components/TableNodeExt/types.ts]: or should we have header columns separated from the data?
its own type and label object ? what do you think?

Comment (minotogna) [reply src/client/components/TableNodeExt/types.ts]: > or should we have header columns separated from the data?
> its own type and label object ? what do you think?

maybe not :P

File: src/client/components/TableNodeExt/CellNodeExt/CellProps.tsx [hunk 11]
```diff
@@ -1,8 +1,22 @@
...
+}
+
+export type CellProps = {
   disabled: boolean
   column: ColumnNodeExt
 }
+
+export type CellMultiselectProps = CellProps & MultiValueProps
+
+export type CellTextProps = CellProps & SingleValueProps
```
Comment (minotogna) [inline src/client/components/TableNodeExt/CellNodeExt/CellProps.tsx]: nice 😊 let's have maybe only two types? ```CellValueSingleProps``` and ```CellValueMultiProps```
what do you think?

File: src/client/components/TableNodeExt/types.ts [hunk 12]
```diff
@@ -7,7 +7,7 @@ export type ColumnNodeExt = {
   type: ColumnNodeExtType
   props: {
     colName: string
-    header: string
+    header: Label
```
Comment (minotogna) [inline src/client/components/TableNodeExt/types.ts]: more flexible if ```header:{label:Label}```

File: src/client/components/TableNodeExt/types.ts (hunk 8 above)
Comment (sorja) [reply src/client/components/TableNodeExt/types.ts]: Introduced `header: { label: Label }` for now

**Review by minotogna — APPROVED** [review]

**Pull request merged**